src/util/http/websocket/UpdateFetcher.h

```diff
@@ -0,0 +1,48 @@
+//   Copyright 2023, University of Freiburg,
+//   Chair of Algorithms and Data Structures.
+//   Author: Robin Textor-Falconi <textorr@informatik.uni-freiburg.de>
+
+#ifndef QLEVER_UPDATEFETCHER_H
+#define QLEVER_UPDATEFETCHER_H
+
+#include <boost/asio/awaitable.hpp>
+#include <boost/asio/use_awaitable.hpp>
+
+#include "util/http/websocket/WebSocketTracker.h"
+
+namespace ad_utility::websocket {
+using StrandType = net::strand<net::any_io_executor>;
+
+/// Class that provides an interface for boost::asio, so a websocket connection
+/// can "asynchronously wait" for an update of a specified query to occur.
+/// There is one instance of this class for every connected websocket.
+class UpdateFetcher {
+  using PayloadType = std::shared_ptr<const std::string>;
+
+  WebSocketTracker& webSocketTracker_;
+  const QueryId& queryId_;
+  StrandType& socketStrand_;
+  std::shared_ptr<QueryToSocketDistributor> distributor_{nullptr};
+  // Counter to ensure sequential processing
+  size_t nextIndex_ = 0;
+
+ public:
+  UpdateFetcher(const QueryId& queryId, WebSocketTracker& webSocketTracker,
+                StrandType& strand)
+      : webSocketTracker_{webSocketTracker},
+        queryId_{queryId},
+        socketStrand_{strand} {}
+
+  /// If an update occurred for the given query since the last time this was
+  /// called this resumes immediately. Otherwise it will wait
```

Comment:
This class should be split up:

The actual `updateFetcher` should only be the `index` and the `sharedPtr<Distributor>`

Then the `WebSocketTracker` should have a function
`awaitable<UpdateFetcher> waitForQueryStart(QueryId)`.

And then this `UpdateFetcher` should either manage its own strand (not passed in), or be not-threadsafe (like the waiting list) and implicitly reuse the strand of the `webSocket` (probably the latter is enough, and similar to what you are currently doing anyway.).
And if that is done, Maybe the remaining functionality goes back into the class that handles a single websocket session, because we are down to one function again here.